pkg/query-service/app/querier/helper.go

```diff
@@ -35,59 +90,88 @@ func (q *querier) runBuilderQuery(
 		preferRPM = q.featureLookUp.CheckFeature(constants.PreferRPM) == nil
 	}
 
-	// TODO: handle other data sources
 	if builderQuery.DataSource == v3.DataSourceLogs {
 		var query string
 		var err error
-		// for ts query with limit replace it as it is already formed
-		if params.CompositeQuery.PanelType == v3.PanelTypeGraph && builderQuery.Limit > 0 && len(builderQuery.GroupBy) > 0 {
-			limitQuery, err := logsV3.PrepareLogsQuery(
-				params.Start,
-				params.End,
-				params.CompositeQuery.QueryType,
-				params.CompositeQuery.PanelType,
-				builderQuery,
-				logsV3.Options{GraphLimitQtype: constants.FirstQueryGraphLimit, PreferRPM: preferRPM},
-			)
+		if _, ok := cacheKeys[queryName]; !ok {
+			query, err = prepareLogsQuery(ctx, params.Start, params.End, builderQuery, params, preferRPM)
 			if err != nil {
-				ch <- channelResult{Err: err, Name: queryName, Query: limitQuery, Series: nil}
+				ch <- channelResult{Err: err, Name: queryName, Query: query, Series: nil}
 				return
 			}
-			placeholderQuery, err := logsV3.PrepareLogsQuery(
-				params.Start,
-				params.End,
-				params.CompositeQuery.QueryType,
-				params.CompositeQuery.PanelType,
-				builderQuery,
-				logsV3.Options{GraphLimitQtype: constants.SecondQueryGraphLimit, PreferRPM: preferRPM},
-			)
+			series, err := q.execClickHouseQuery(ctx, query)
+			ch <- channelResult{Err: err, Name: queryName, Query: query, Series: series}
+			return
```

Comment:
With the return here, will the result ever get cached? There is no cache key to begin with there will never be one because we return.

Reply:
This logic is for queries is where caching is not required.

Reply:
cacheKeys are generated by the GenerateKeys, so it will always be there for queries that needs to use cache.